lib/pbench/server/database/models/server_config.py

```diff
@@ -0,0 +1,389 @@
+
+        Args:
+            readonly: Caller requires only read access to the server API.
+
+        Returns:
+            None if the server is enabled. If the server is disabled for the
+            requested access, the entire JSON value is returned and should be
+            reported to a caller.
+        """
+        state = __class__.get(key=ServerConfigOptions.SERVER_STATE.key)
+        if state:
+            value = state.value
+            status = value[STATE_STATUS_KEY]
+            if status == "disabled" or status == "readonly" and not readonly:
+                return value
+        return None
+
+    @staticmethod
+    def get_all() -> JSONOBJECT:
+        """
+        Return all server config settings as a JSON object
+        """
+        configs = Database.db_session.query(ServerConfig).all()
+        db = {c.key: c.value for c in configs}
+        return {k: db[k] if k in db else __class__._default(k) for k in __class__.KEYS}
+
+    def __str__(self) -> str:
+        """
+        Return a string representation of the key-value pair
+
+        Returns:
+            string representation
+        """
+        return f"{self.key}: {self.value!r}"
+
+    def _decode(self, exception: IntegrityError) -> Exception:
+        """
+        Decode a SQLAlchemy IntegrityError to look for a recognizable UNIQUE
+        or NOT NULL constraint violation. Return the original exception if
+        it doesn't match.
+
+        Args:
+            exception: An IntegrityError to decode
+
+        Returns:
+            a more specific exception, or the original if decoding fails
+        """
+        # Postgres engine returns (code, message) but sqlite3 engine only
+        # returns (message); so always take the last element.
+        cause = exception.orig.args[-1]
+        if cause.find("UNIQUE constraint") != -1:
+            return ServerConfigDuplicate(self.key, cause)
+        elif cause.find("NOT NULL constraint") != -1:
+            return ServerConfigNullKey(self.key, cause)
+        return exception
+
+    def add(self):
+        """
+        Add the ServerConfig object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except IntegrityError as e:
+            Database.db_session.rollback()
+            raise self._decode(e)
+        except Exception as e:
+            Database.db_session.rollback()
+            raise ServerConfigSqlError("adding", self.key, str(e))
+
+    def update(self):
+        """
+        Update the database row with the modified version of the
+        ServerConfig object.
+        """
+        try:
+            Database.db_session.commit()
+        except IntegrityError as e:
+            Database.db_session.rollback()
+            raise self._decode(e)
+        except Exception as e:
+            Database.db_session.rollback()
+            raise ServerConfigSqlError("updating", self.key, str(e))
```

Comment:
Again, cut-and-paste. In datasets and metadata I didn't even bother decoding the `IntegrityError`, mostly because the template had less constrained input (files read from disk) and it wasn't safe to presume we wouldn't hit SQL column constraints at update. I'm not so sure that's necessary here, and I probably left it more because the decoding is kinda cool.

And although I've definitely done `except Exception` with `isinstance` or `type` decoding, I really prefer not to, because it's harder to read the logic. In this case, if I keep both `except` clauses, I'd prefer to keep the few lines of implementation separate as well. If I convince myself I don't need `_decode`, then it won't matter...